### ISSUE TITLE
ICE: resolve_trait_associated_const: unexpected vtable type

### ISSUE THREAD

**Issue by oli-obk** [rust-lang/rust#29924]
the following code ices on nightly:

``` rust
#![feature(const_fn, associated_consts)]

trait Trait {
    const N: usize;
}

impl Trait {
    const fn n() -> usize { Self::N }
}
```

It should give an error instead noting that you can't do `Self::N`.

**Comment by bltavares** [comment]
Triaging: this issue is already being tracked on https://github.com/steveklabnik/glacier

Reference: steveklabnik/glacier#144

**Comment by niconii** [comment]
Still happens as of `rustc 1.16.0-nightly (4ecc85beb 2016-12-28)`.

**Comment by aiwenar** [comment]
As of `rustc 1.17.0-nightly (ea7a6486a 2017-02-04)` this can also be hit with `<Trait as Trait>::CONST`

```rust
#![feature(associated_consts)]

trait Trait {
  const N: usize;
}

fn n() -> usize {
  <Trait as Trait>::N
}
```

**Comment by Mark-Simulacrum** [comment]
The first example doesn't compile due to the trait not being object safe and the last example also doesn't compile today. Neither ICEs; E-needstest.

**Comment by kvark** [comment]
@Mark-Simulacrum 
Here is a new case:
```rust
pub enum Foo {
    A
}
pub trait Food {
    const FOO: Foo;
}

pub enum A {}
impl Food for A {
    const FOO: Foo = Foo::A;
}

pub trait Bard {
    type Foo: Food;
    fn get_foo() -> Foo { Self::Foo::FOO }
}
```
Error on Rust-1.20 ([playground](https://play.rust-lang.org/?gist=ee45b0b9b8bfd02a6741e981309ef4f8&version=stable)):
> error: internal compiler error: /checkout/src/librustc_const_eval/eval.rs:529: resolve_trait_associated_const: unexpected vtable type

Note: this is sort of a blocker for us to get use of the associated consts.

**Comment by oli-obk** [comment]
@kvark this is already fixed on beta/nightly

**Comment by kvark** [comment]
@oli-obj great, but what about stable? No hot patches planned?

**Comment by Mark-Simulacrum** [comment]
We generally do not backport patches to stable unless they are critical bug fixes (for which I don't think I can give a concrete definition). We should have a stable out with the fix in 6 weeks, though.
